scripts/test/Muon/loadfile_presenter_test.py

```diff
@@ -0,0 +1,510 @@
+        self.view.set_file_edit = mock.Mock()
+        self.mock_model_to_load_workspaces([[1]], [1234])
+        self.mock_user_input_text("C:/dir1/file1.nxs")
+
+        self.presenter.handle_file_changed_by_user()
+        self.Runner(self.presenter._load_thread)
+
+        self.assertEqual(self.model.loaded_filenames, ["C:/dir1/file1.nxs"])
+        self.assertEqual(self.model.loaded_workspaces, [[1]])
+        self.assertEqual(self.model.loaded_runs, [1234])
+
+        self.view.set_file_edit.assert_called_once_with("C:/dir1/file1.nxs", mock.ANY)
+
+    @run_test_with_and_without_threading
+    def test_that_if_invalid_file_selected_in_browser_view_does_not_change(self):
+        self.mock_browse_button_to_return_files(["not_a_file"])
+        self.mock_model_to_load_workspaces([[1]], [1234])
+
+        self.view.set_file_edit = mock.Mock()
+        self.view.reset_edit_to_cached_value = mock.Mock()
+
+        self.presenter.on_browse_button_clicked()
+        self.Runner(self.presenter._load_thread)
+
+        self.model.load_workspace_from_filename = mock.Mock(side_effect=self.load_failure)
+
+        cc = self.view.set_file_edit.call_count
+        self.presenter.on_browse_button_clicked()
+        self.Runner(self.presenter._load_thread)
+
+        self.assertEqual(self.view.set_file_edit.call_count, cc)
+        self.assertEqual(self.view.reset_edit_to_cached_value.call_count, 0)
+
+    @run_test_with_and_without_threading
+    def test_that_view_reverts_to_previous_text_if_users_supplies_invalid_text(self):
+        self.load_workspaces_into_model_and_view_from_browse([[1]], [1234], ["C:/dir1/EMU0001234.nxs"])
+
+        invalid_user_input = ["some random text", "1+1=2", "..."]
+
+        call_count = self.view.reset_edit_to_cached_value.call_count
+        for invalid_text in invalid_user_input:
+            call_count += 1
+            self.view.get_file_edit_text = mock.Mock(return_value=invalid_text)
+
+            self.presenter.handle_file_changed_by_user()
+            self.Runner(self.presenter._load_thread)
+
+            self.assertEqual(self.view.reset_edit_to_cached_value.call_count, call_count)
+
+    @run_test_with_and_without_threading
+    def test_that_model_and_interface_revert_to_previous_values_if_load_fails_from_browse(self):
+        self.load_workspaces_into_model_and_view_from_browse([[1]], [1234], ["C:/dir1/EMU0001234.nxs"])
+
+        self.model.load_workspace_from_filename = mock.Mock(side_effect=self.load_failure)
+
+        self.presenter.on_browse_button_clicked()
+        self.Runner(self.presenter._load_thread)
+
+        self.assertEqual(self.model.loaded_filenames, ["C:/dir1/EMU0001234.nxs"])
+        self.assertEqual(self.model.loaded_workspaces, [[1]])
+        self.assertEqual(self.model.loaded_runs, [1234])
+
+        self.assertEqual(self.view.reset_edit_to_cached_value.call_count, 0)
+        # TODO : add a check of the actual text
```

Comment:
is this not line 278?

Reply:
No, line 278 is testing the model. My comment is referring to the view. Thanks for spotting this TODO though I will fix it.